src/Native/Windows/clrcompression/zlib-intel/deflate.c

```diff
@@ -165,88 +151,62 @@ local const config configuration_table[10] = {
+#define RANK(f) (((f) * 2) - ((f) > 4 ? 9 : 0))
 
 
 /* ===========================================================================
- * Insert string str in the dictionary and set match_head to the previous head
- * of the hash chain (the most recent string with same hash key). Return
- * the previous length of the hash chain.
- * If this file is compiled with -DFASTEST, the compression level is forced
- * to 1, and no hash chains are maintained.
- * IN  assertion: all calls to to INSERT_STRING are made with consecutive
- *    input characters and the first MIN_MATCH bytes of str are valid
- *    (except for the last MIN_MATCH-1 bytes of the input file).
+ * Initialize the hash table (avoiding 64K overflow for 16 bit systems).
+ * prev[] will be initialized on the fly.
  */
-#ifdef USE_SSE4_2_CRC_HASH
-local inline Pos insert_string_sse(deflate_state *z_const s, z_const Pos str)
-{
-    Pos ret;
-    unsigned *ip, val, h = 0;
-
-    ip = (unsigned *)&s->window[str];
-    val = *ip;
-
-    if (s->level >= 6)
-        val &= 0xFFFFFF;
-
-#ifndef _MSC_VER
-    __asm__ __volatile__ (
-        "crc32 %1,%0\n\t"
-    : "+r" (h)
-    : "r" (val)
-    );
-#else
-    h = _mm_crc32_u32(h, val);
-#endif
-
-    ret = s->head[h & s->hash_mask];
-    s->head[h & s->hash_mask] = str;
-    s->prev[str & s->w_mask] = ret;
-    return ret;
-}
-#endif
+#define CLEAR_HASH(s) \
+    do { \
+        s->head[s->hash_size-1] = NIL; \
+        zmemzero((Bytef *)s->head, \
+                 (unsigned)(s->hash_size-1)*sizeof(*s->head)); \
+    } while (0)
 
-local inline Pos insert_string_c(deflate_state *z_const s, z_const Pos str)
+/* ===========================================================================
+ * Slide the hash table when sliding the window down (could be avoided with 32
+ * bit values at the expense of memory usage). We slide even when level == 0 to
+ * keep the hash table consistent if we switch back to level > 0 later.
+ */
+local void slide_hash_c(s)
+    deflate_state *s;
 {
-    Pos ret;
+    unsigned n, m;
+    Posf *p;
+    uInt wsize = s->w_size;
 
-    UPDATE_HASH(s, s->ins_h, str);
-#ifdef FASTEST
-    ret = s->head[s->ins_h];
-#else
-    ret = s->prev[str & s->w_mask] = s->head[s->ins_h];
+    n = s->hash_size;
+    p = &s->head[n];
+    do {
+        m = *--p;
+        *p = (Pos)(m >= wsize ? m - wsize : NIL);
+    } while (--n);
+    n = wsize;
+#ifndef FASTEST
+    p = &s->prev[n];
+    do {
+        m = *--p;
+        *p = (Pos)(m >= wsize ? m - wsize : NIL);
+        /* If n is not on any hash chain, prev[n] is garbage but
+         * its value will never be used.
+         */
+    } while (--n);
 #endif
-    s->head[s->ins_h] = str;
-
-    return ret;
 }
 
-local inline Pos insert_string(deflate_state *z_const s, z_const Pos str)
+local void slide_hash(deflate_state *s)
 {
-#ifdef USE_SSE4_2_CRC_HASH
-    if (x86_cpu_has_sse42)
-        return insert_string_sse(s, str);
+#ifdef USE_SSE_SLIDE
+    if (x86_cpu_has_sse2)
```

Comment:
Do we need this, since CoreCLR requires SSE2 for both x86 and x64? Is this to support other runtimes using CoreFX that might not have the same requirement?